python/tchannel/sync.py

```diff
@@ -0,0 +1,140 @@
+        response = client.request(
+            hostport='localhost:4040',
+            service='HelloService',
+        ).send(
+            'hello', None, json.dumps({"name": "World"})
+        )
+
+    """
+
+    def __init__(self, name, process_name=None, known_peers=None, trace=False):
+        """Initialize a new TChannelClient.
+
+        :param process_name:
+            Name of the calling process. Used for logging purposes only.
+        """
+        self.async_client = async.TChannel(
+            name,
+            hostport=glossary.EPHEMERAL_HOSTPORT,
+            process_name=process_name,
+            known_peers=known_peers,
+            trace=trace
+        )
+        self.threadloop = ThreadLoop()
+        self.threadloop.start()
+
+    def request(self, *args, **kwargs):
+        """Initiate a new request to a peer.
+
+        :param hostport:
+            If specified, requests will be sent to the specific host.
+            Otherwise, a known peer will be picked at random.
+        :param service:
+            Name of the service being called. Defaults to an empty string.
+        :param service_threshold:
+            If ``hostport`` was not specified, this specifies the score
+            threshold at or below which peers will be ignored.
+        :returns SyncClientOperation:
+            An object with a ``send(arg1, arg2, arg3)`` operation.
+        """
+        operation = self.async_client.request(*args, **kwargs)
+        operation = SyncClientOperation(operation, self.threadloop)
+
+        return operation
+
+
+class SyncClientOperation(object):
+    """Allows making client operation requests synchronously.
+
+    Composes a tchannel.
+
+    This object acts like tchannel.TChannelClientOperation, but instead
+    uses a temporary ioloop to make the request synchronously.
+    """
+
+    def __init__(self, operation, threadloop):
+        assert operation, "operation is required"
+        assert threadloop, "threadloop.ThreadLoop is required"
+        self.operation = operation
+        self.threadloop = threadloop
+
+    def send(self, arg1, arg2, arg3):
+        """Send the given triple over the wire.
+
+        :param arg1:
+            String containing the contents of arg1. If None, an empty string
+            is used.
+        :param arg2:
+            String containing the contents of arg2. If None, an empty string
+            is used.
+        :param arg3:
+            String containing the contents of arg3. If None, an empty string
+            is used.
+        :return concurrent.futures.Future:
+            Future response from the peer.
+        """
+        arg1 = arg1 or ''
+        arg2 = arg2 or ''
+        arg3 = arg3 or ''
+
+        @gen.coroutine
+        def make_request():
+
+            response = yield self.operation.send(arg1, arg2, arg3)
+
+            header = yield response.get_header()
+            body = yield response.get_body()
+
+            result = Response(header, body)
```

Comment:
I'm not totally sure if we should be converting the TChannel to the stripped down Response here... perhaps it should look more like the actual Response?